### PR TITLE
ANON-205 - Anon dns introduction

### PR DESCRIPTION
**Pull request by yumirkov**

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -19152,7 +19152,7 @@ Changes in version 0.2.9.4-alpha - 2016-10-17
 
 Changes in version 0.2.9.3-alpha - 2016-09-23
   Tor 0.2.9.3-alpha adds improved support for entities that want to make
-  high-performance services available through the Tor .onion mechanism
+  high-performance services available through the Tor .anon mechanism
```
Comment (0xneuratic) [inline ChangeLog:19155]: Replace Tor with Anyone

File: ChangeLog [hunk 2]
```diff
@@ -34465,7 +34465,7 @@ Changes in version 0.2.0.19-alpha - 2008-02-09
   o Code simplifications and refactoring:
     - Remove some needless generality from cpuworker code, for improved
       type-safety.
-    - Stop overloading the circuit_t.onionskin field for both "onionskin
+    - Stop overloading the circuit_t.anonskin field for both "onionskin
```
Comment (0xneuratic) [inline ChangeLog:34468]: would keep it onionskin unless the variable itself has changed (e.g., is it anonrc?)

**Comment by yumirkov** [comment]
Client should have `anons` file in data dir with content with the following format:

```
xyz.anon x6f7aejcj55bpkml4w4s36ej5oatozux5ko3ly7ycj76hfccbcnolxad.anon
```

**Review by Saundr21 — APPROVED** [review]

File: src/core/or/connection_edge.c [hunk 3]
```diff
@@ -1764,6 +1771,46 @@ parse_extended_hostname(char *address, hostname_type_t *type_out)
   return false;
 }
 
+bool lookup_anon_dns_mapping(const char *anon_address, char *onion_address_out) {
```
Comment (kanshi) [inline src/core/or/connection_edge.c:1774]: Validate input params

```
if (!anon_address || !onion_address_out) {
    return false;
}
```

File: src/core/or/connection_edge.c [hunk 4]
```diff
@@ -1764,6 +1771,46 @@ parse_extended_hostname(char *address, hostname_type_t *type_out)
...
+  // Check if the file exists using `file_status`
+  char *dns_fname = get_datadir_fname("anons");
+  file_status_t terms_status = file_status(dns_fname);
+  if (terms_status != FN_FILE) {
+    log_notice(LD_APP,"DNS mapping file 'anons' is not found in data dir.");
+    return false;
+  }
+
+  // Read the entire file content using `read_file_to_str`
+  file_content = read_file_to_str(dns_fname, 0, NULL);
```
Comment (kanshi) [inline src/core/or/connection_edge.c:1786]: Add some reasonable file size limit.

`file_content = read_file_to_str(dns_fname, MAX_FILE_SIZE, NULL);`

File: src/core/or/connection_edge.c [hunk 5]
```diff
@@ -1764,6 +1771,46 @@ parse_extended_hostname(char *address, hostname_type_t *type_out)
   return false;
 }
 
+bool lookup_anon_dns_mapping(const char *anon_address, char *onion_address_out) {
+  char *file_content = NULL;
+
+  // Check if the file exists using `file_status`
+  char *dns_fname = get_datadir_fname("anons");
```
Comment (kanshi) [inline src/core/or/connection_edge.c:1778]: Might be useful to sanity check the output

File: src/core/or/connection_edge.c (hunk 3 above)
Comment (kanshi) [reply src/core/or/connection_edge.c:1774]: Also, should verify the size of the out_buffer?

File: src/core/or/connection_edge.c [hunk 6]
```diff
@@ -1764,6 +1771,46 @@ parse_extended_hostname(char *address, hostname_type_t *type_out)
...
+    return false;
+  }
+
+  // Process the file content line by line
+  char *line = strtok(file_content, "\n");
+  while (line != NULL) {
+    char anon[256];
+    char onion[HS_SERVICE_ADDR_LEN_BASE32 + 6 + 1];
+    // Parse each line into anon and onion components
+    if (sscanf(line, "%255s %62s", anon, onion) == 2) {
```
Comment (kanshi) [inline src/core/or/connection_edge.c:1798]: Should we validate the content of the anons file? At least sanity checks if operators will be able to manually edit those?

**Review by kanshi — CHANGES_REQUESTED** [review]

File: src/core/or/connection_edge.c (hunk 6 above)
Comment (Saundr21) [reply src/core/or/connection_edge.c:1798]: + to kanshi's comment here. Think best to validate content of anons file.

File: src/core/or/connection_edge.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (yumirkov) [reply src/core/or/connection_edge.c:1778]: No. We decided not to create a default file for the first iteration.

Comment (yumirkov) [reply src/core/or/connection_edge.c:1793]: I did not get it. This is the existing function for data-dir file management

Comment (kanshi) [reply src/core/or/connection_edge.c:1793]: I see the file_status is just mapping results. Nvm.

**Review by kanshi — APPROVED** [review]

**Pull request merged**